corehq/apps/reports/views.py

```diff
@@ -479,65 +484,80 @@ def touch_saved_reports_views(user, domain):
-    except ValueError:
-        # invalidly formatted date input
-        return HttpResponseBadRequest()
-
-    date_range = POST.get('date_range')
-    if date_range == 'last7':
-        POST['days'] = 7
-    elif date_range == 'last30':
-        POST['days'] = 30
-    elif POST.get('days'):
-        POST['days'] = int(POST['days'])
+class AddSavedReportConfigView(View):
+    name = 'add_report_config'
 
-    exclude_filters = ['startdate', 'enddate']
-    for field in exclude_filters:
-        POST['filters'].pop(field, None)
+    @method_decorator(login_and_domain_required)
+    def post(self, request, domain, *args, **kwargs):
+        self.domain = domain
 
-    config = ReportConfig.get_or_create(POST.get('_id', None))
+        if not self.saved_report_config_form.is_valid():
+            return HttpResponseBadRequest()
 
-    if config.owner_id:
-        # in case a user maliciously tries to edit another user's config
-        assert config.owner_id == user_id
-    else:
-        config.domain = domain
-        config.owner_id = user_id
+        update_config_data = deepcopy(self.saved_report_config_form.cleaned_data)
+        del update_config_data['_id']
+        update_config_data.update({
+            'filters': self.filters,
+        })
+        for field in self.config.properties().keys():
+            if field in update_config_data:
+                setattr(self.config, field, update_config_data[field])
+
+        # remove start and end date if the date range is "last xx days"
+        if (
+            self.saved_report_config_form.cleaned_data['days']
+            or self.saved_report_config_form.cleaned_data['date_range'] == 'lastmonth'
+        ):
+            if "start_date" in self.config:
+                delattr(self.config, "start_date")
+            if "end_date" in self.config:
+                delattr(self.config, "end_date")
+
+        self.config.save()
+        ReportsTab.clear_dropdown_cache(request, self.domain)
+        touch_saved_reports_views(request.couch_user, self.domain)
+
+        return json_response(self.config)
+
+    @property
+    @memoized
+    def config(self):
+        config = ReportConfig.get_or_create(
+            self.saved_report_config_form.cleaned_data['_id']
+        )
+        if config.owner_id:
+            # in case a user maliciously tries to edit another user's config
+            assert config.owner_id == self.user_id
+        else:
+            config.domain = self.domain
+            config.owner_id = self.user_id
+        return config
+
+    @property
+    @memoized
+    def saved_report_config_form(self):
+        return SavedReportConfigForm(
+            self.domain,
+            self.user_id,
+            self.post_data
+        )
 
-    for field in config.properties().keys():
-        if field in POST:
-            setattr(config, field, POST[field])
+    @property
+    def filters(self):
+        filters = deepcopy(self.post_data.get('filters', {}))
```

Comment:
Looks like a shallow copy would work here as well. They tend to be much cheaper.